pkgs/tools/typesetting/tex/texlive/combine.nix

```
@@ -73,22 +58,48 @@ in (buildEnv {
   paths = lib.catAttrs "outPath" pkgList.bin ++ [ doc ];
   pathsToLink = [
     "/"
-    "/bin" # ensure these are writeable directories
+    "/bin"
+    "/share/texmf-config"
+    "/share/texmf-var/web2c" # ensure these are writeable directories
   ];
 
-  nativeBuildInputs = [ makeWrapper libfaketime perl bin.texlinks ];
-  buildInputs = pkgList.extraInputs;
+  nativeBuildInputs = [
+    makeWrapper
+    libfaketime
+    (lib.last tl.texlive-scripts.pkgs) # fmtutil, mktexlsr, updmap
+    (lib.last tl.texlive-scripts-extra.pkgs) # texlinks
+  ];
 
   # This is set primarily to help find-tarballs.nix to do its job
   passthru.packages = pkgList.all;
 
-  postBuild = ''
+  postBuild =
+    # environment variables (note: only export the ones that are used in the wrappers)
+  ''
     TEXMFDIST="${texmf}"
     export PATH="$out/bin:$PATH"
-    export PERL5LIB="$TEXMFDIST/scripts/texlive:${bin.core.out}/share/texmf-dist/scripts/texlive"
     TEXMFSYSCONFIG="$out/share/texmf-config"
     TEXMFSYSVAR="$out/share/texmf-var"
     export TEXMFCNF="$TEXMFSYSVAR/web2c"
+  '' +
+    # wrap executables with required env vars as early as possible
+    # 1. we want texlive.combine to use the wrapped binaries, to catch bugs
+    # 2. we only wrap genuine binaries, and not e.g. links generated by texlinks
+  ''
+    BASH_LOADABLES_PATH=${bash}/lib/bash
+    enable -f realpath realpath
+    declare -i wrapCount=0
+    for link in "$out"/bin/*; do
+      target=$(realpath "$link")
```

This is problematic because `--inherit-argv0` does not work for scripts with shebang (at least on linux), so some of the wrapped scripts will behave as the wrong program.

---

I have resorted to using hard links everywhere instead of symbolic links (or even copies – I figured hard links within the same derivation are ok). Does `--inherit-argv0` have any side effects, or can we leave it there?

---

for interpreted scripts with shebang, the value of `argv[0]` is discarded, setting it doesn't do any harm. It's better to keep `--inherit-argv0`  since some of the compiled binaries might just check whether `argv[0]` is equal to the expected program name (and not also cover the case where it's a path), just to be safe.

hardlinking inside a derivation works, but when the store path is serialized as a NAR, they will become separate files with identical content, that will eventually be hardlinked again when store optimisation kicks in. Keeping symlinks for binaries and only using copies for scripts would also work, but is slightly more error-prone.

Is there any reason not to use symlinks and only resolve a single symlink instead of fully canonicalizing the target path for the wrapper (as was done previously)?: 

```suggestion
      target=$(readlink "$link") # resolves one symlink
```

---

I didn't like the extra layer of indirection (in the worst case, format link => wrapper => bin container link => executable). Not terrible, but there is a simple solution after all: check in `combine` if the basename looks right, and if not, use `readlink` instead of `realpath`. This is enough to get exactly what we want, in all scenarios: we create symlinks instead of copies in the binary containers, and we stat exactly as many links as needed for each binary, and no more.

Also, since there was a lot of duplicate code around aliases, I introduced an extra attribute set `binaliases` that does all the work (assert that it is a subset of `binfiles`, create symlinks, check for existence, canonicalise, ...). This is starting to look good.

---

This should be ok now.